File: c++/22_Generate_Parentheses.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <string>

using namespace std;

// reference: 
// http://bangbingsyb.blogspot.com/2014/11/leetcode-generate-parentheses.html

class Solution {
public:
    vector<string> generateParenthesis(int n) {
    	vector<string> allCombo;
    	string combo;
    	backTracking(n, 0, 0, combo, allCombo);
    	return allCombo;

    }
    void backTracking(int n, int left, int right, string combo, vector<string> &allCombo) {
    	
    	if (left == n && right == n) {
    		allCombo.push_back(combo);
    		return;
    	}
    	if (left < n) {
    		//combo += '(';
    		combo.push_back('(');
    		backTracking(n, left + 1, right, combo, allCombo);
    		combo.pop_back();
    	}
    	if (right < left) {
    		combo.push_back(')');
    		backTracking(n, left, right + 1, combo, allCombo);
    		combo.pop_back();
    	}
    }    
};

int main() {
    Solution s;
    vector<string> res;
    res = s.generateParenthesis(3);
    for (int i = 0; i < res.size(); i++)
    	cout << res[i] << " ";
    cout << endl;
    
}

File: python/22_Generate_Parentheses.py
#!/usr/bin/python
import sys
import os


'''
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> allComb;
        string comb;
        findParenthesis(n, 0, 0, comb, allComb);
        return allComb;
    }
    
    void findParenthesis(int n, int nleft, int nright, string &comb, vector<string> &allComb) {
        if(nleft==n && nright==n) {
            allComb.push_back(comb);
            return;
        }
        
        if(nleft<n) {
            comb.push_back('(');
            findParenthesis(n, nleft+1, nright, comb, allComb);
            comb.pop_back();
        }
        
        if(nright<nleft) {
            comb.push_back(')');
            findParenthesis(n, nleft, nright+1, comb, allComb);
            comb.pop_back();
        }
    }
};
'''

class Solution(object):
    def generateParenthesis(self, n):
        allComb = []
        comb = ""
        self.backTracking(n, 0, 0, comb, allComb)
        return allComb
    
    def backTracking(self, n, left, right, comb, allComb):
    	print allComb
    	if n == left and n == right:
    		allComb.append(comb)
    	if left < n:
    		self.backTracking(n, left + 1, right, comb + "(", allComb)

    	if right < left:
    		self.backTracking(n, left, right + 1, comb + ")", allComb)
    	
    

def main():
    s = Solution()
    print s.generateParenthesis(3)

if __name__ == "__main__":
    main()
